---
title: 'LR analysis using MultiNicheNet: For single Cell Transcriptomic data'
author: "Gayani Anandagoda"
date: "`r Sys.Date()`"
output: html_document
params: 
  treatment: "default"
  cell_idents: ["default"]
  batches: "default"
  covariates: "default"
  
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# LR analysis using MultiNicheNet can be done step by step in 6 different steps or run the wrapper script at the end

```{r load packages, message=FALSE}
library(Seurat)
library(SeuratObject)
library(SingleCellExperiment)
library(dplyr)
library(ggplot2)

library(nichenetr)
library(multinichenetr)
```

```{r load data}
# Load Seurat object
# This Seurat object is the final output from the Data_preprocessing.Rmd script
# This Seurat object contains samples only with MHCII expression levels High and Low
# cell types also have been edited and available as "edited_cell_type" in meta data

filtered_scRNA_obj_all = readRDS("/run/user/1356082/gvfs/smb-share:server=group3.ad.helsinki.fi,share=h345/afarkkilab/Projects/NKI/Single_Cell_LR_Annalysis/Input/launonen_cancercell_allcells_allannot_Filtered_MHCII_HIgh_and_low.RDS")

```

```{r parameters}

organism = "human"

treatment = params$treatment  # taking the sample IDs of the chemo naive samples

cell_idents = c(params$cell_idents)

# Define cell types to include in the analysis

# Cell types availble in "edited_cell_type" in meta data of "launonen_cancercell_allcells_allannot_Filtered.RDS"
# [1] "FOXP3_CD4_Treg"         "Stroma"                 "CD8_T_cell"             "CD4_T_cell"            
# [5] "NK"                     "Tumor"                  "B_cell"                 "macrophages_CD163_high"
# [9] "DC"                     "Other_immune"           "macrophages_ITGAX_high"

min_cells = 5  # cell type filtering : determines which cell types are sufficiently present
min_sample_prop = 0.50 # Gene filtering : determines which genes are sufficiently expressed in each present cell type
fraction_cutoff = 0.05 # Gene filtering : fraction of cells in the sample with non-zero expression values

ligand_activity_down = FALSE # focus on up-regulated ligands

empirical_pval = FALSE
logFC_threshold = 0.75 
p_val_threshold = 0.05
p_val_adj = TRUE 


top_n_target = 250
verbose = TRUE
cores_system = 8
```


```{r contrasts table}

sample_id = "sample"  
group_id = "MHCIIGroups_bulk"
celltype_id = "edited_cell_type" 
batches = ifelse(params$batches == "default", NA,params$batches) # batch correction
covariates = ifelse(params$covariates == "default", NA,params$covariates) # add patient Ids if there are paired samples

# Set contrasts
contrasts_oi <- c("'High-Low','Low-High'")

# Create a contrast table
contrast_tbl <- tibble(contrast = c("High-Low","Low-High"), 
                       group = c("High", "Low"))
```

```{r output directory}
#  define directory
output_dir = "/run/user/1356082/gvfs/smb-share:server=group3.ad.helsinki.fi,share=h345/afarkkilab/Projects/NKI/Single_Cell_LR_Annalysis/output/"

```

```{r load lr networks}

options(timeout = 120)

if(organism == "human"){
  
  lr_network_all = 
    readRDS(url(
      "https://zenodo.org/record/10229222/files/lr_network_human_allInfo_30112033.rds"
      )) %>% 
    mutate(
      ligand = convert_alias_to_symbols(ligand, organism = organism), 
      receptor = convert_alias_to_symbols(receptor, organism = organism))
  
  lr_network_all = lr_network_all  %>% 
    mutate(ligand = make.names(ligand), receptor = make.names(receptor)) 
  
  lr_network = lr_network_all %>% 
    distinct(ligand, receptor)
  
  ligand_target_matrix = readRDS(url(
    "https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"
    ))
  
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  
  lr_network = lr_network %>% filter(ligand %in% colnames(ligand_target_matrix))
  ligand_target_matrix = ligand_target_matrix[, lr_network$ligand %>% unique()]
  
} else if(organism == "mouse"){
  
  lr_network_all = readRDS(url(
    "https://zenodo.org/record/10229222/files/lr_network_mouse_allInfo_30112033.rds"
    )) %>% 
    mutate(
      ligand = convert_alias_to_symbols(ligand, organism = organism), 
      receptor = convert_alias_to_symbols(receptor, organism = organism))
  
  lr_network_all = lr_network_all  %>% 
    mutate(ligand = make.names(ligand), receptor = make.names(receptor)) 
  lr_network = lr_network_all %>% 
    distinct(ligand, receptor)
  
  ligand_target_matrix = readRDS(url(
    "https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final_mouse.rds"
    ))
  
  colnames(ligand_target_matrix) = colnames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  rownames(ligand_target_matrix) = rownames(ligand_target_matrix) %>% 
    convert_alias_to_symbols(organism = organism) %>% make.names()
  
  lr_network = lr_network %>% filter(ligand %in% colnames(ligand_target_matrix))
  ligand_target_matrix = ligand_target_matrix[, lr_network$ligand %>% unique()]
  
}

```


```{r subset seurat object}
if(treatment == "all"){
  
  filtered_scRNA_obj_all = filtered_scRNA_obj_all 
} else {
  
  cn_samples = filtered_scRNA_obj_all@meta.data %>% filter(treatment_stage == treatment) %>% pull(sample) %>% unique()

  filtered_scRNA_obj_all = subset(filtered_scRNA_obj_all, subset = sample %in% cn_samples)
  
}
```


```{r prepare data, message=FALSE,results='hide'}
######################### Preparing the data and generating diagnostic plots ##############

# Convert Seurat to SCE
sce <- Seurat::as.SingleCellExperiment(filtered_scRNA_obj_all, assay = "RNA")

# make sure that gene symbols used in the expression data are updated
sce = alias_to_symbol_SCE(sce, "human") %>% makenames_SCE()

# Define sender and receiver cell types
senders_oi <- SummarizedExperiment::colData(sce)[, celltype_id] %>% unique() %>% .[.%in% cell_idents]

receivers_oi <- SummarizedExperiment::colData(sce)[, celltype_id] %>% unique() %>% .[.%in% cell_idents]

sce = sce[, SummarizedExperiment::colData(sce)[,celltype_id] %in% 
            c(senders_oi, receivers_oi)
          ]
```


```{r diagnostic plots}

abundance_expression_info <- get_abundance_info(sce = sce, 
                                                           sample_id = sample_id, 
                                                           group_id = group_id, 
                                                           celltype_id = celltype_id, 
                                                           min_cells = min_cells, 
                                                           senders_oi = senders_oi, 
                                                           receivers_oi = receivers_oi, 
                                                           #lr_network = lr_network, 
                                                           batches = batches)


# diagnostic plots
# To check whether each cell type have enough number of cells in each sample

file_name_abundance_expression_info = paste0(output_dir,"abundance_expression_info_",treatment,".RDS") # file name

saveRDS(abundance_expression_info, file = file_name_abundance_expression_info)

abundance_expression_info$abund_plot_sample
abundance_expression_info$abund_plot_group
abundance_expression_info$abund_barplot
```

```{r  Cell-type filtering}

################################ Analysis ###############################

# Step 01. Cell-type filtering

abundance_df_summarized = abundance_expression_info$abundance_data %>% 
  mutate(keep = as.logical(keep)) %>% 
  group_by(group_id, celltype_id) %>% 
  summarise(samples_present = sum((keep)))

celltypes_absent_one_condition = abundance_df_summarized %>% 
  filter(samples_present == 0) %>% pull(celltype_id) %>% unique() 
# find truly condition-specific cell types by searching for cell types 
# truely absent in at least one condition

celltypes_present_one_condition = abundance_df_summarized %>% 
  filter(samples_present >= 2) %>% pull(celltype_id) %>% unique() 
# require presence in at least 2 samples of one group so 
# it is really present in at least one condition

condition_specific_celltypes = intersect(
  celltypes_absent_one_condition, 
  celltypes_present_one_condition)

total_nr_conditions = SummarizedExperiment::colData(sce)[,group_id] %>% 
  unique() %>% length() 

absent_celltypes = abundance_df_summarized %>% 
  filter(samples_present < 2) %>% 
  group_by(celltype_id) %>% 
  dplyr::count() %>% 
  filter(n == total_nr_conditions) %>% 
  pull(celltype_id)
  
print("condition-specific celltypes:")
## [1] "condition-specific celltypes:"
print(condition_specific_celltypes)
## character(0)
  
print("absent celltypes:")
## [1] "absent celltypes:"
print(absent_celltypes)
## character(0)

# condition-specific cell types can be filtered out if you as a user do not want to run the alternative workflow for condition-specific cell types in the optional step 8 of the core MultiNicheNet analysis. check !!
```



```{r Gene Filtering}

############## Step 02. Gene Filtering ############

frq_list = get_frac_exprs(
  sce = sce, 
  sample_id = sample_id, celltype_id =  celltype_id, group_id = group_id, 
  batches = batches, 
  min_cells = min_cells, 
  fraction_cutoff = fraction_cutoff, 
  min_sample_prop = min_sample_prop)

# Now only keep genes that are expressed by at least one cell type:

genes_oi = frq_list$expressed_df %>% 
  filter(expressed == TRUE) %>% pull(gene) %>% unique() 
sce = sce[genes_oi, ]

```



```{r Pseudobulk expression calculation}

################ Step 03. Pseudobulk expression calculation ####################

abundance_expression_info = process_abundance_expression_info(
  sce = sce, 
  sample_id = sample_id, group_id = group_id, celltype_id = celltype_id, 
  min_cells = min_cells, 
  senders_oi = senders_oi, receivers_oi = receivers_oi, 
  lr_network = lr_network, 
  batches = batches, 
  frq_list = frq_list, 
  abundance_info = abundance_expression_info)


# Normalized pseudobulk expression values per gene/celltype/sample can be inspected by:

#abundance_expression_info$celltype_info$pb_df %>% head()
#abundance_expression_info$celltype_info$pb_df_group %>% head()
#abundance_expression_info$sender_receiver_info$pb_df %>% head()
#abundance_expression_info$sender_receiver_info$pb_df_group %>% head()
```

```{r DE Analysis}

## Step 04. Differential expression (DE) analysis: #####

DE_info = get_DE_info(
  sce = sce, 
  sample_id = sample_id, group_id = group_id, celltype_id = celltype_id, 
  batches = batches, covariates = covariates, 
  contrasts_oi = contrasts_oi, 
  min_cells = min_cells, 
  expressed_df = frq_list$expressed_df)

#DE_info$celltype_de$de_output_tidy %>% head()


if(empirical_pval == TRUE){
  DE_info_emp = get_empirical_pvals(DE_info$celltype_de$de_output_tidy)
  celltype_de = DE_info_emp$de_output_tidy_emp %>% select(-p_val, -p_adj) %>% 
    rename(p_val = p_emp, p_adj = p_adj_emp)
} else {
  celltype_de = DE_info$celltype_de$de_output_tidy
} 


# Combine DE information for ligand-senders and receptors-receivers

sender_receiver_de = multinichenetr::combine_sender_receiver_de(
  sender_de = celltype_de,
  receiver_de = celltype_de,
  senders_oi = senders_oi,
  receivers_oi = receivers_oi,
  lr_network = lr_network
)

#head(sender_receiver_de)

# plot to check the p value distribution
DE_info$hist_pvals
```

```{r Ligand activity prediction}

###### Step 05. Ligand activity prediction #####

geneset_assessment = contrast_tbl$contrast %>% 
  lapply(
    process_geneset_data, 
    celltype_de, logFC_threshold, p_val_adj, p_val_threshold
  ) %>% 
  bind_rows() 



# Perform the ligand activity analysis and ligand-target inference
# increase the number of scores to run the code faster

n.cores = min(cores_system, celltype_de$cluster_id %>% unique() %>% length()) 


ligand_activities_targets_DEgenes = suppressMessages(suppressWarnings(
  get_ligand_activities_targets_DEgenes(
    receiver_de = celltype_de,
    receivers_oi = intersect(receivers_oi, celltype_de$cluster_id %>% unique()),
    ligand_target_matrix = ligand_target_matrix,
    logFC_threshold = logFC_threshold,
    p_val_threshold = p_val_threshold,
    p_val_adj = p_val_adj,
    top_n_target = top_n_target,
    verbose = verbose, 
    n.cores = n.cores
  )
))

ligand_activities_targets_DEgenes$ligand_activities %>% head(5)
```



```{r saveRDS}
file_name_ligand_activities = paste0(output_dir,"ligand_activities_targets_DEgenes_",treatment,".RDS") # file name

saveRDS(ligand_activities_targets_DEgenes, file = file_name_ligand_activities )
```



```{r Prioritization}


##### Step 06. Prioritization: rank cell-cell communication patterns through multi-criteria prioritization
sender_receiver_tbl = sender_receiver_de %>% distinct(sender, receiver)

metadata_combined = SummarizedExperiment::colData(sce) %>% tibble::as_tibble()

if(!is.na(batches)){
  grouping_tbl = metadata_combined[,c(sample_id, group_id, batches)] %>% 
    tibble::as_tibble() %>% distinct()
  colnames(grouping_tbl) = c("sample","group",batches)
} else {
  grouping_tbl = metadata_combined[,c(sample_id, group_id)] %>% 
    tibble::as_tibble() %>% distinct()
  colnames(grouping_tbl) = c("sample","group")
}

prioritization_tables = suppressMessages(multinichenetr::generate_prioritization_tables(
    sender_receiver_info = abundance_expression_info$sender_receiver_info,
    sender_receiver_de = sender_receiver_de,
    ligand_activities_targets_DEgenes = ligand_activities_targets_DEgenes,
    contrast_tbl = contrast_tbl,
    sender_receiver_tbl = sender_receiver_tbl,
    grouping_tbl = grouping_tbl,
    scenario = "regular", # all prioritization criteria will be weighted equally
    fraction_cutoff = fraction_cutoff, 
    abundance_data_receiver = abundance_expression_info$abundance_data_receiver,
    abundance_data_sender = abundance_expression_info$abundance_data_sender,
    ligand_activity_down = ligand_activity_down
  ))

# This table gives the final prioritization score of each interaction, and the values of the individual prioritization criteria.

prioritization_tables$group_prioritization_tbl %>% head(5)
```


```{r save table}
# SAVE THE "prioritization_tables" TABLE AS A R OBJECT


file_name_table = paste0(output_dir,"prioritization_tables_",treatment,".RDS") # file name

saveRDS(prioritization_tables, file = file_name_table)
```

################### Visualization
# To Generate the plots see Plot_generation.Rmd: The script uses the outputs saved in the prioritization_tables 

```{r eval=FALSE}
# Wrapper script
# instead of running the code chunks from step 1 to 6 we can run this wrapper script


multinichenet_output = multi_nichenet_analysis(
  sce = sce, 
  celltype_id = celltype_id, sample_id = sample_id, group_id = group_id, 
  batches = batches, covariates = covariates, 
  lr_network = lr_network, ligand_target_matrix = ligand_target_matrix, 
  contrasts_oi = contrasts_oi, contrast_tbl = contrast_tbl, 
  senders_oi = senders_oi, receivers_oi = receivers_oi,
  min_cells = min_cells, 
  fraction_cutoff = fraction_cutoff, 
  min_sample_prop = min_sample_prop,
  scenario = scenario, 
  ligand_activity_down = ligand_activity_down,
  logFC_threshold = logFC_threshold, 
  p_val_threshold = p_val_threshold, 
  p_val_adj = p_val_adj, 
  empirical_pval = empirical_pval, 
  top_n_target = top_n_target, 
  n.cores = n.cores, 
  verbose = TRUE
  )


```


```{r}
sessionInfo()
```

