---
title: "R Notebook"
output: html_notebook
---

```{r}
setwd("C:/Users/fernpere/")
library(ggplot2)
library(tidyverse)
library(DESeq2)
library(sigQC)
library(survival)
library(survminer)
library(gridExtra)
```

#Reading bulk RNAseq from TCGA
```{r}
TCGA.bulk.counts <- read.table(file="C:/Users/fernpere/Downloads/RNAseq_TCGA/BulkRNA_TCGA_OVA_augmented_star_gene_counts.tsv",
                                sep="\t", header=TRUE)
```

#Reading HRD status shared in Perez-Villatoro et al 2022 and clinical data from TCGA
```{r}
input.folder <- "D:/users/fperez/NKI_TMAs_AF/TCGA_PRISM_RNAdeconvolution/"


clin.dat <- read.csv(file = paste0(input.folder,"TCGA-CDR-SupplementalTableS1_stage-survival.csv"),
                      header = TRUE, row.names = 1)

out.put.folder <- "D:/users/fperez/NKI_TMAs_AF/Analysis_results/05_ML_validations/TCGA_RNA/"
dir.create(out.put.folder)

HRD.tab <- read.table(file=paste0(input.folder,"Categories_Konstantinopoulos_HR_HGSC_germline.csv"),
                      sep=",", header=TRUE)


#Stratify samples according to tumoral molecular profiles
HRD.tab$Molecular.profile <- "HRP"
HRD.tab$Molecular.profile[HRD.tab$categories == "CCNE1 amplification"] <- "CCNE1amp"
HRD.tab$Molecular.profile[HRD.tab$HRDsum >= 54] <- "HRD"
HRD.tab$Molecular.profile[grep("BRCA", HRD.tab$categories)] <- "BRCAloss"
HRD.tab.cat <- HRD.tab[,c("Sample","Molecular.profile")]

# Structuring clinical data
clin.dat <- clin.dat[clin.dat$type == "OV",]
clin.dat$PFI.time <- as.numeric(clin.dat$PFI.time)/30.4
clin.dat$OS.time <- as.numeric(clin.dat$OS.time)/30.4
clin.dat$OS <- as.numeric(clin.dat$OS)
clin.dat$PFI <- as.numeric(clin.dat$PFI)
clin.dat$age <- as.numeric(clin.dat$age_at_initial_pathologic_diagnosis)

clin.dat$age.bin <- NA
clin.dat$age.bin[clin.dat$age < 40] = 0
clin.dat$age.bin[clin.dat$age >= 40 & clin.dat$age < 50] = 1
clin.dat$age.bin[clin.dat$age >= 50 & clin.dat$age < 60] = 2
clin.dat$age.bin[clin.dat$age >= 60 & clin.dat$age < 70] = 3
clin.dat$age.bin[clin.dat$age >= 70 & clin.dat$age < 80] = 4
clin.dat$age.bin[clin.dat$age >= 80] = 5

#Selecting only high patients with high grade tumors 
clin.dat <- clin.dat[clin.dat$histological_grade %in% c("G3","G4"),]

#Ignoring low clinical stages
clin.dat <- clin.dat[!clin.dat$clinical_stage %in% c("[Not Available]", "Stage IA","Stage IB", "Stage IC",
                                                         "Stage IIA","Stage IIB","Stage IIC"),]
clin.dat$bcr_patient_barcode <- gsub("-", ".", clin.dat$bcr_patient_barcode)
clin.dat

HRD.tab.cat$Sample <- gsub("-", ".", HRD.tab.cat$Sample)
```


#Normalizing RNA.count.expression using DESeq
```{r}
#Vector with samples, the variable group is random, will not be really used
samples <- data.frame(Sample = colnames(TCGA.bulk.counts)[-1],
                      group=sample(c("Low","High"),length(colnames(TCGA.bulk.counts)[-1]), replace = TRUE))


#Preparing input for DESeq
m.counts <- as.matrix(TCGA.bulk.counts[,-1])
row.names(m.counts) <- TCGA.bulk.counts[,1]

deseq <- DESeqDataSetFromMatrix(countData = round(m.counts),
                                colData = samples,
                                design = as.formula("~group"))

#Selecting only genes with at least 1 counts per group
deseq <- deseq[rowSums(counts(deseq)) > 1, ]
#Stimating size factor with DESeq
deseq <- estimateSizeFactors(deseq)

#Selecting only rows with at least expression of 1
m.counts <- m.counts[rowSums(m.counts) > 1,]

#Normalizing the expression
norm.counts.per.sample <- as.data.frame(t(m.counts) * deseq$sizeFactor)

rm(TCGA.bulk.counts)
rm(m.counts)
rm(deseq)
norm.counts.per.sample
```

#Merging normalized counts with clinical data
```{r}
norm.counts.clin.prev <- merge(clin.dat, norm.counts.per.sample, by.x="bcr_patient_barcode", by.y="row.names")

norm.counts.clin <- merge(norm.counts.clin.prev, HRD.tab.cat, by.x="bcr_patient_barcode", by.y="Sample")
rm(norm.counts.clin.prev)

```


#Preparing dataset for signature dectection
```{r}
#These are the genes differentially expressed using bulk-RNA
tcga.DE.genes <- read.table(file="C:/Users/fernpere/Downloads/RNAseq_TCGA/Overexpressed_genes.csv",
                          sep=",", header=TRUE)

#Detecting the first and last gene column in dataset
gene.cols <- which(colnames(norm.counts.clin) %in% c("TSPAN6", "AP006621.6"))

mRNA_expr_matrix = list()
aux <- norm.counts.clin[,-1]
row.names(aux) <- norm.counts.clin[,1]
gene.cols <- gene.cols-1
aux <- aux[,gene.cols[1]:gene.cols[2]]
mRNA_expr_matrix[["TCGA_bulk"]] =  t(aux)

gene_sigs_list = list()

gene_list_names = c("MHCIIgenes", "MHCIgenes","NanostringDE","Intersected","TCGAbulkDE")
list_of_genes=list(c("HLA-DPA1","HLA-DPA2","HLA-DPA3","HLA-DPB1","HLA-DPB2","HLA-DQA1","HLA-DQA2","HLA-DQB1",
                 "HLA-DQB2","HLA-DQB3","HLA-DRA","HLA-DRB1","HLA-DRB2","HLA_DRB3","HLA_DRB4","HLA-DRB5",
                 "HLA-DRB6","HLA-DRB7","HLA-DRB8","HLA-DRB9","CD74"),
                 c("HLA-A","HLA-B","HLA-C","HLA-E","HLA-F","HLA-G","HLA-H","HLA-J","HLA-K","HLA-L",
                   "HLA-N","HLA-P","HLA-S","HLA-T","HLA-U","HLA-V","HLA-W","HLA-X","HLA-Y","HLA-Z"),
                c("CD74", "CXCL9", "CXCL10", "CXCL11","IDO1", "ADAMDEC1", "SLAMF7", "CCL5", "CD3D"),
                c("CCL5", "CD74", "CXCL9"),
                tcga.DE.genes$gene)

for (i in 1:length(gene_list_names)){
  genes <- intersect(row.names(mRNA_expr_matrix$TCGA_bulk), list_of_genes[[i]])
  gene_sigs_list[[gene_list_names[i]]] = genes
}

rm(aux)
gene_sigs_list
```

#Running program for MHCII gene signature
#The result is stored in the out_dir
```{r}
out.put.dir = 'D:/users/fperez/NKI_TMAs_AF/Analysis_results/05_ML_validations/TCGA_RNA/SignaturesQC_bulk/'

make_all_plots(gene_sigs_list = gene_sigs_list,
               mRNA_expr_matrix = mRNA_expr_matrix,
               doNegativeControl = FALSE,
               out_dir = out.put.dir,
               showResults = FALSE)
```



#Reading signatures and merging results with dataset
```{r}
standarization_tables <- list.files(paste0(out.put.dir,"/standardisation_tables"),
                                    pattern = "_TCGA_bulk.txt", full.names = TRUE)

norm.counts.clin2 = norm.counts.clin

for (f in (standarization_tables)){
  signature.name <- strsplit(basename(f), "_")[[1]][3]
  sig <- read.table(file=f, sep="\t", header=TRUE)
  colnames(sig) <- c(paste0(signature.name,".sig"), paste0(signature.name,".sig.scaled"))
  norm.counts.clin2 <- cbind(norm.counts.clin2, sig[norm.counts.clin2$bcr_patient_barcode,])
}
```



#Stratiifying samples according to MHCII expression and performing survival analysis
```{r}
total.genes.to.test <- c("CD74", "CXCL9", "CXCL10", "CXCL11",
                   "IDO1", "ADAMDEC1", "SLAMF7", "CCL5", "CD3D")

hrp.genes.to.test <- c("CRABP1", "CD2", "SLAMF7", "ADAMDEC1", "CD3E", "CD27", "CTLA4", "IFNG",
                      "CD3D", "CXCL9", "CXCL10", "CXCL11", "IDO1")

signatures <- paste0(gene_list_names, ".sig")

genes.to.test <- unique(c(hrp.genes.to.test, total.genes.to.test, signatures))

out.put.folder2 <- "D:/users/fperez/NKI_TMAs_AF/Analysis_results/05_ML_validations/TCGA_RNA/Bulk_RNA/"

dir.create(out.put.folder2)


formulas.names <- c("OS","PFI")
formulas.labs <- c("OS probability","PFI probability")
xlim.max.val <- c(150,120)
dat <- norm.counts.clin2

formulas.cox <- c('Surv(OS.time, OS) ~ age.bin + Molecular.profile',
              'Surv(PFI.time, PFI)~ age.bin + Molecular.profile')

formulas.kapplan <- c('Surv(OS.time, OS) ~  Expression',
              'Surv(PFI.time, PFI)~ Expression')

for (gene in genes.to.test){
  for (i in 1:length(formulas.kapplan)){
    #For cox model data is stratified by tiles 
    dat$Expression <- ntile(dat[,gene],3)
    colnames(dat)[colnames(dat) == "Expression"] = paste0(gene,"exp")
  
    model <- coxph(as.formula(paste0(formulas.cox[i]," + ", gene,"exp")), data=dat)
    p1 <- ggforest(model, data = dat,  main=formulas.labs[i])
    print(p1)
    ggsave(p1, file=paste0(out.put.folder2, "IMMUNE_Cox_",gene,"_age_",formulas.names[i],".png"),
           width = 15, height = 10, units = "cm")
    
    #For kapplan data is stratified by median expression 
    dat$Expression <- ifelse(dat[,gene] > median(dat[,gene]), "High","Low")
    
    fit <- survfit(as.formula(formulas.kapplan[i]), data = dat)
    logrank = surv_pvalue(fit, dat)
    pval=logrank$pval.txt
    p <- ggsurvplot(fit, data = dat, risk.table = TRUE, p.val=TRUE, conf.int = TRUE,
                    palette = c("#ff5a36","#4d4dff"),
                   legend.labs=c(paste0(gene,".High"), paste0(gene,".Low")),
                   tables.y.text = FALSE, legend.title="",
                   xlim=c(0,xlim.max.val[i]), break.time.by=30, ylab=formulas.labs[i])
    p$plot <- p$plot + theme(legend.key.width = unit(3, "line"),
                             plot.margin = margin(t = 0,  r = 2, b = 0, l = 2, unit = "cm"),
                             legend.key.height = unit(1.5, "line"))
    p$plot <- p$plot + ggplot2::annotate("text", x = 90, y = 0.85, label =pval, size = 5)
    p$table <- p$table + theme(plot.margin = margin(t = 0,  r = 2, b = 5, l = 2, unit = "cm"))
    plot.to.save <- list(p$plot, p$table)
    ggsave(file=paste0(out.put.folder2,"IMMUNE_Kapplan_",gene,"_",formulas.names[i],".png"),
            arrangeGrob(grobs = plot.to.save, ncol = 1),
          width = 13, height = 16, units = "cm")
  }
}
```
